pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -474,6 +474,43 @@ func (m *kubeGenericRuntimeManager) executePreStopHook(pod *v1.Pod, containerID
 	return int64(metav1.Now().Sub(start.Time).Seconds())
 }
 
+// executeRetryUntilSuccess runs the Retry-Until lifecycle hooks if applicable and returns the duration it takes.
+func (m *kubeGenericRuntimeManager) executeRetryUntilSuccessHook(pod *v1.Pod, containerID kubecontainer.ContainerID, containerSpec *v1.Container, gracePeriod int64) int64 {
+	glog.V(3).Infof("Running RetryUntilSuccess hook for container %q", containerID.String())
+
+	start := metav1.Now()
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		defer utilruntime.HandleCrash()
+		if msg, err := m.runner.Run(containerID, pod, containerSpec, containerSpec.Lifecycle.PreStop); err != nil {
+			glog.Errorf("retryUntilSuccess hook for container %q failed: %v", containerSpec.Name, err)
+			m.generateContainerEvent(containerID, v1.EventTypeWarning, events.FailedPreStopHook, msg)
+			ticker := time.NewTicker(5 * time.Second)
+                        rangeNumber := ((gracePeriod / 5) - 1)
+			for i:=int64(0); i< rangeNumber; i++ {
```

Comment:
@brendandburns Hi , Imagine if Prestop will never succeed. The But long time timeout will trigger the select in the code, and then the container will be forcibly removed, but go func () would never stop because the container would not be able to perform the Prestop hook, so I used the For Range loop here. If there is a better plan, please let me know